omniscidb/StringDictionary/StringDictionary.cpp

```diff
@@ -1408,3 +1416,315 @@ size_t StringDictionary::buildDictionaryTranslationMap(
+                                          uint8_t* encoded_vec) const;
+template size_t StringDictionary::getBulk(const std::vector<std::string>& string_vec,
+                                          uint16_t* encoded_vec) const;
+template size_t StringDictionary::getBulk(const std::vector<std::string>& string_vec,
+                                          int32_t* encoded_vec) const;
+
+template <class T, class String>
+size_t StringDictionary::getBulk(const std::vector<String>& string_vec,
+                                 T* encoded_vec,
+                                 const int64_t generation) const {
+  std::atomic<size_t> num_strings_not_found;
+  tbb::parallel_for(tbb::blocked_range<size_t>(0, string_vec.size()),
+                    [&](const tbb::blocked_range<size_t>& r) {
+                      for (size_t i = r.begin(); i != r.end(); ++i) {
+                        const auto& str = string_vec[i];
+                        if (str.empty()) {
+                          encoded_vec[i] = inline_int_null_value<T>();
+                        } else {
+                          if (str.size() > StringDictionary::MAX_STRLEN) {
+                            legacy::throw_string_too_long_error(str, dict_ref_);
+                          }
+
+                          const auto hash = hash_string(str);
+                          const auto string_id = getIdOfStringImpl(hash, str);
+                          if (string_id == StringDictionary::INVALID_STR_ID ||
+                              string_id > int32_t(strings.size())) {
+                            encoded_vec[i] = StringDictionary::INVALID_STR_ID;
+                            num_strings_not_found++;
+                          }
+                          encoded_vec[i] = string_id;
+                        }
+                      }
+                    });
+
+  return num_strings_not_found.load();
+}
+
+template size_t StringDictionary::getBulk(const std::vector<std::string>& string_vec,
+                                          uint8_t* encoded_vec,
+                                          const int64_t generation) const;
+template size_t StringDictionary::getBulk(const std::vector<std::string>& string_vec,
+                                          uint16_t* encoded_vec,
+                                          const int64_t generation) const;
+template size_t StringDictionary::getBulk(const std::vector<std::string>& string_vec,
+                                          int32_t* encoded_vec,
+                                          const int64_t generation) const;
+
+template <class T, class String>
+void StringDictionary::getOrAddBulk(const std::vector<String>& string_vec,
+                                    T* output_string_ids) {
+  mapd_lock_guard<mapd_shared_mutex> write_lock(rw_mutex_);
+
+  // compute hashes
+  auto hashes = std::make_unique<uint32_t[]>(string_vec.size());
+  tbb::parallel_for(tbb::blocked_range<size_t>(0, string_vec.size()),
```

Comment:
The legacy version used a grain size to avoid very small tasks. I think it's reasonable to put some limits here. We don't want to create a task per each string.

Reply:
Thanks for the tips! I experimented with a few different options following [this guide](https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Controlling_Chunking_os.html) and something in the range of 25k seemed to be the best, though not by very much. It's possible taxi is too noisy and we need a dedicated microbenchmark to tune this. I think 25k should be fine for now, though.